[dv/doc] security verification in cip_lib (#10692)
lowRISC/opentitan: hw/dv/sv/cip_lib/doc/index.md

side, which is the UVM environment extended from CIP library. The diagram
lists some common items that need to be instantiated in `tb`
and set into `uvm_config_db` for the testbench to work.

## Security Verification in cip_lib
CIP contains reusable security verification components, sequences and function coverage.

> This section describes the details of them and the steps to enabling them.

Reviewer: I am actually not sure which one is grammarly correct - 
```suggestion
This section describes the details of them and the steps to enable them.
```

TODO, add methodology docs for security verification and add the link here

cip_lib imports [sec_cm_pkg](https://github.com/lowRISC/opentitan/tree/master/hw/dv/sv/sec_cm), which automatically locates all the common countermeasure primitives and binds an interface to each of them. In the cib_base_vseq, it injects fault to each of these primitives and verifies that the fault will lead to a fatal alert.
If the block uses common security countermeasure primitives (prim_count, prim_sparse_fsm_flop, prim_double_lfsr), users can enable this sequence to fully verify them via following steps.
The details of the sequences can be found in testplans - [sec_cm_count_testplan](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/dvsim/testplans/sec_cm_count_testplan.hjson), [sec_cm_fsm_testplan](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/dvsim/testplans/sec_cm_fsm_testplan.hjson) and [sec_cm_double_lfsr_testplan](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/dvsim/testplans/sec_cm_double_lfsr_testplan.hjson).

1. Import the applicable sec_cm testplans.
If more checks or sequences are needed, add another testpoint in the block testplan.

> For example, when the fault is detected by countermeasure, the following operation won’t be executed.

Reviewer 2: ```suggestion
For example, when the fault is detected by countermeasure, some subsequent operation won’t be executed.
```

2. Import [sec_cm_tests](https://github.com/lowRISC/opentitan/blob/master/hw/dv/tools/dvsim/tests/sec_cm_tests.hjson) in sim_cfg.hjson file, as well as add applicable sec_cm bind files for `sim_tops`.
The `ip_sec_cm` test will be added and all common countermeasure primitives will be verified in this test.

```
sim_tops: ["ip_ctrl_bind", "ip_ctrl_cov_bind",
           // only add the corresponding bind file if DUT has the primitive
           "sec_cm_prim_sparse_fsm_flop_bind",
           "sec_cm_prim_count_bind",
           "sec_cm_prim_double_lfsr_bind"]
```

3. Import sec_cm_pkg in the env_pkg
```systemverilog
package ip_env_pkg;
  import uvm_pkg::*;
  import sec_cm_pkg::*;
  …
```

4. Set alert name for countermeasure if the alert name is different from the default name - “fatal_fault”.
```systemverilog
class ip_env_cfg extends cip_base_env_cfg #(.RAL_T(ip_reg_block));
  virtual function void initialize(bit [31:0] csr_base_addr = '1);
    super.initialize(csr_base_addr);
    sec_cm_alert_name = "fatal_check_error";
```

5. Override the `check_sec_cm_fi_resp` task in ip_common_vseq to add additional sequences and checks after fault injection.

> This is the example from keymgr, in which CSR `fault_status` will be updated according to the location of the fault and the operation after fault inject will lead design to enter `StInvalid` state.

Reviewer 2: ```suggestion
This is an example from keymgr, in which CSR `fault_status` will be updated according to the location of the fault and the operation after fault inject will lead design to enter `StInvalid` state.
```
